src/coprocessor/dag/rpn_expr/types/expr_builder.rs

```diff
@@ -278,6 +296,28 @@ where
     // Map pb func to `RpnFnMeta`.
     let func = fn_mapper(tree_node.get_sig(), tree_node.get_children())?;
     let args = tree_node.take_children().into_vec();
+
+    // Only Int/Real/Duration/Decimal/Bytes/Json will be decoded
+    let datums = datum::decode(&mut tree_node.get_val())?;
+    for d in datums.iter() {
+        if let Datum::Null = d {
+            return Err(box_err!("push down datum should be NOT NULL"));
+        }
+    }
+    let implicit_args = datums
```

Comment:
why not simply combine them to one loop? I believe compiler cannot optimize two loops + two branches into one loop + one branch..

Reply:
The original intention was to avoid allocating a vector if the datums contain `Null`, anyway, you are right because of error should be rarely.